### PR TITLE
Create file over blob endpoint.

### PR DESCRIPTION
**Pull request by anmolanmol1234**
This PR has the handling to make a PutBlob call if the prefix mode is Blob. A separate API has been added in the client class which does PutBlob call for creating files. Store has the intelligence to call the needed API based on prefix mode.

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 1]
```diff
@@ -311,15 +319,24 @@ public FSDataOutputStream create(final Path f, final FsPermission permission, fi
 
     statIncrement(CALL_CREATE);
     trailingPeriodCheck(f);
+    TracingContext tracingContext = new TracingContext(clientCorrelationId,
+            fileSystemId, FSOperationType.CREATE, overwrite, tracingHeaderFormat, listener);
 
     Path qualifiedPath = makeQualified(f);
+    boolean fileOverwrite = overwrite;
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java:346]: Add comments that this is needed for create idempotency when overwrite is false.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 2]
```diff
@@ -262,6 +262,14 @@ public void registerListener(Listener listener1) {
     listener = listener1;
   }
 
+  public boolean isNamespaceEnabled() {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Unable to see callers for this and getPrefixMode. is this needed anywhere ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 3]
```diff
@@ -1491,7 +1508,7 @@ public boolean failed() {
   }
 
   @VisibleForTesting
-  AzureBlobFileSystemStore getAbfsStore() {
+  public AzureBlobFileSystemStore getAbfsStore() {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Can this be avoided by bringing the test class to same namespace ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 4]
```diff
@@ -559,18 +568,28 @@ public OutputStream createFile(final Path path,
...
       } else {
-        op = client.createPath(relativePath, true,
-            overwrite,
-            isNamespaceEnabled ? getOctalNotation(permission) : null,
-            isNamespaceEnabled ? getOctalNotation(umask) : null,
-            isAppendBlob,
-            null,
-            tracingContext);
-
+        if (getPrefixMode() == PrefixMode.BLOB) {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:800]: This should be Blob && !isNamespace, as the changes are not scoped for HNS + blob

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 5]
```diff
@@ -635,9 +660,14 @@ private AbfsRestOperation conditionalCreateOverwriteFile(final String relativePa
             .getResponseHeader(HttpHeaderConfigurations.ETAG);
 
         try {
-          // overwrite only if eTag matches with the file properties fetched befpre
-          op = client.createPath(relativePath, true, true, permission, umask,
-              isAppendBlob, eTag, tracingContext);
+          // overwrite only if eTag matches with the file properties fetched before.
+          if (getPrefixMode() == PrefixMode.BLOB) {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:896]: When AbfsClient make the decision to fork to blob or DFS API based on prefix mode ?

**Review by snvijaya — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 2 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Usage is in test classes, isNamespaceEnabled is used in class ITestAbfsFileSystemContractEtag and getPrefixMode is used in class ITestAzureBlobFileSystemLease.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 4 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:800]: PrefixMode would be BLOB only for FNS accounts. This variable is set in the initialize method of AzureBlobFileSystem. So we don't need explicit check for isNamespaceEnabled.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 5 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:896]: Have added separate API in AbfsClient for create file. So, it doesn't need fork based on prefix mode.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 3 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Fixed this.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 1 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java:346]: Done.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 6]
```diff
@@ -262,6 +262,16 @@ public void registerListener(Listener listener1) {
     listener = listener1;
   }
 
+  @VisibleForTesting
+  boolean isNamespaceEnabled() {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: These were removed in the append PR, same clean up is needed here too.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 7]
```diff
@@ -600,10 +613,10 @@ public void deleteFilesystem(TracingContext tracingContext)
     }
   }
 
-  public OutputStream createFile(final Path path,
+  public OutputStream createFile(final Path path, final boolean isFile,
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Rename to isDirMarkerFile and reverse all checks.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 8]
```diff
@@ -600,10 +613,10 @@ public void deleteFilesystem(TracingContext tracingContext)
...
       final FileSystem.Statistics statistics, final boolean overwrite,
       final FsPermission permission, final FsPermission umask,
-      TracingContext tracingContext) throws IOException {
+      TracingContext tracingContext, HashMap<String, String> metadata) throws IOException {
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:762]: if its known its a directory marker, why would metadata be needed ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 9]
```diff
@@ -712,9 +776,14 @@ private AbfsRestOperation conditionalCreateOverwriteFile(final String relativePa
...
-          op = client.createPath(relativePath, true, true, permission, umask,
-              isAppendBlob, eTag, tracingContext);
+          // overwrite only if eTag matches with the file properties fetched before.
+          if (getPrefixMode() == PrefixMode.BLOB) {
+            op = client.createPathBlob(relativePath, true, true, metadata,
+                    eTag, tracingContext);
+          } else {
+            op = client.createPath(relativePath, true, true, permission, umask,
+                    isAppendBlob, eTag, tracingContext);
+          }
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:901]: There are 3 places that need to createPath. abstracting creation (line 780 to 786) and returning op will reduce code duplication.

**Review by snvijaya — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 7 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Removed this param dependency

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 6 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Fixed

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 8 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:762]: Metadata is needed for sending metadata during put blob

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 9 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:901]: done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 10]
```diff
@@ -300,6 +304,19 @@ protected CompletableFuture<FSDataInputStream> openFileWithOptions(
             open(path, Optional.of(parameters.getOptions())));
   }
 
+  // This handling makes sure that if request to create file for an existing directory or subpath
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: nit: javadocs for the method.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 11]
```diff
@@ -600,10 +611,33 @@ public void deleteFilesystem(TracingContext tracingContext)
...
-  public OutputStream createFile(final Path path,
-      final FileSystem.Statistics statistics, final boolean overwrite,
+  /**
+   * Checks if we are creating a normal blob or markerFile.
+   * @param metadata takes metadata as param.
+   * @return true or false.
+   */
+  private boolean checkIsBlobOrMarker(HashMap<String, String> metadata) {
+    return metadata != null && metadata.containsKey(X_MS_META_HDI_ISFOLDER) &&
+            metadata.get(X_MS_META_HDI_ISFOLDER).equalsIgnoreCase(TRUE);
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: nit:
```
metdata != null && TRUE.equalsIgnoreCase(metdatada.get(X_MS_META_HDI_ISFOLDER)
```

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 12]
```diff
@@ -665,6 +699,27 @@ public OutputStream createFile(final Path path,
...
+   * Returns true if the path is a directory.
+   * @param path path to check for file or directory.
+   * @param tracingContext tracingContext.
+   * @return true or false.
+   */
+  boolean checkIsDirectory(Path path, TracingContext tracingContext) throws IOException {
+   BlobProperty blobProperty;
+    try {
+      blobProperty = getBlobProperty(path, tracingContext);
+    } catch (AzureBlobFileSystemException ex) {
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: lets have `AbfsRestOperationException`: reduce casting, if condition.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 13]
```diff
@@ -223,7 +226,8 @@ List<AbfsHttpHeader> createDefaultHeaders() {
     final List<AbfsHttpHeader> requestHeaders = new ArrayList<AbfsHttpHeader>();
     requestHeaders.add(new AbfsHttpHeader(X_MS_VERSION, xMsVersion));
     requestHeaders.add(new AbfsHttpHeader(ACCEPT, APPLICATION_JSON
-            + COMMA + SINGLE_WHITE_SPACE + APPLICATION_OCTET_STREAM));
+            + COMMA + SINGLE_WHITE_SPACE + APPLICATION_OCTET_STREAM
+            + COMMA + SINGLE_WHITE_SPACE + APPLICATION_XML));
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:231]: why changing for all apis? can it be intrusive to other APIs? If not, we can let it be.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 14]
```diff
@@ -428,6 +429,59 @@ public AbfsRestOperation createPath(final String path, final boolean isFile, fin
...
+    if (!overwrite) {
+      requestHeaders.add(new AbfsHttpHeader(IF_NONE_MATCH, AbfsHttpConstants.STAR));
+    }
+    if (eTag != null && !eTag.isEmpty()) {
+      requestHeaders.add(new AbfsHttpHeader(HttpHeaderConfigurations.IF_MATCH, eTag));
+    }
+
+    final AbfsUriQueryBuilder abfsUriQueryBuilder = createDefaultUriQueryBuilder();
+
+    String operation = SASTokenProvider.CREATE_FILE_OPERATION;
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:448]: lets have another enum. In future. reason, in case something changes in future for putBlob, Dev will have a way to give steps for the putBlob in their SASTokenProvider implementation.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 15]
```diff
@@ -428,6 +429,59 @@ public AbfsRestOperation createPath(final String path, final boolean isFile, fin
...
+    try {
+      op.execute(tracingContext);
+    } catch (AzureBlobFileSystemException ex) {
+      // If we have no HTTP response, throw the original exception.
+      if (!op.hasResult()) {
+        throw ex;
+      }
+      if (!isFile && op.getResult().getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        if (metadata != null && metadata.containsKey(IS_FOLDER_METADATA_KEY)
+                && metadata.get(IS_FOLDER_METADATA_KEY).equals(TRUE)) {
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:474]: nit:
```
metadata != null && TRUE.equalsIgnoreCase(metadata.get(IS_FOLDER_METADATA_KEY)
```

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+        final AzureBlobFileSystem fs = getFileSystem();
+        final Configuration configuration = new Configuration();
+        configuration.addResource(TEST_CONFIGURATION_FILE_NAME);
+        AbfsClient abfsClient = getClient(fs);
+
+        AbfsConfiguration abfsConfiguration = new AbfsConfiguration(configuration,
+                configuration.get(FS_AZURE_ABFS_ACCOUNT_NAME));
+
+        // Gets the client.
+        AbfsClient testClient = Mockito.spy(TestAbfsClient.createTestClientFromCurrentContext(
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java]: why we need a spy?

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+                    AbfsRestOperation op = testClient.createPathBlob(testPath.toUri().getPath(),
+                            true, false, null, null, tracingContext);
+                    if (op.getResult().getStatusCode() == HTTP_CONFLICT) {
+                        throw new IOException("BlobAlreadyExists");
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException(e); // pass the exception through as-is
+                }
+            }));
+        }
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java]: 1. This need not enforce parallel hits.
2. If two request gets parallelize such that two request go to 2 diff machine at same time. Can server maintain isolation. Can two blob get created?
3. test requirement can be achieved from sequential run as well. Why parallelism is required.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+        int exceptionCaught = 0;
+        for (Future<?> future : futures) {
+            try {
+                future.get(); // wait for the task to complete and handle any exceptions thrown by the lambda expression
+            } catch (ExecutionException e) {
+                exceptionCaught++;
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    intercept(RuntimeException.class, () -> {
+                        throw (RuntimeException) cause; // re-throw the RuntimeException
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java]: why is this required.

**Review by saxenapranav — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java (hunk 14 above)
Comment (snvijaya) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:448]: SASTokenProvider is a public one that clients will have to incorporate in their implementation, creating more enums for same purpose wont be desirable. Also, didnt catch what it means for - Dev will have a way to give steps for the putBlob
Can you explain

**Review by snvijaya — APPROVED** [review]
Once @saxenapranav's comments are addressed, lets commit to dev.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java (hunk 13 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:231]: This just tells what all formats are acceptable, does not affect other API's

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 12 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 11 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 10 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 19]
```diff
@@ -428,6 +429,59 @@ public AbfsRestOperation createPath(final String path, final boolean isFile, fin
...
 
+  public AbfsRestOperation createPathBlob(final String path, final boolean isFile, final boolean overwrite,
+                                          final HashMap<String, String> metadata,
+                                          final String eTag,
+                                          TracingContext tracingContext)
+          throws AzureBlobFileSystemException {
+    final List<AbfsHttpHeader> requestHeaders = createDefaultHeaders();
+    if (isFile) {
+      addCustomerProvidedKeyHeaders(requestHeaders);
+    }
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: lets not header so that, CPK is not supported.

Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: Removed the header

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java (hunk 16 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java]: Addressed

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java (hunk 17 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestBlobOperation.java]: Removed

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 20]
```diff
@@ -605,6 +614,29 @@ public void deleteFilesystem(TracingContext tracingContext)
...
+   */
+  private boolean checkIsBlobOrMarker(HashMap<String, String> metadata) {
+    return metadata != null && TRUE.equalsIgnoreCase(metadata.get(X_MS_META_HDI_ISFOLDER));
+  }
+
+  private AbfsRestOperation createPath(final String path, final boolean isFile, final boolean overwrite,
+                                      final String permission, final String umask,
+                                      final boolean isAppendBlob, final String eTag,
+                                      TracingContext tracingContext) throws AzureBlobFileSystemException {
+    return client.createPath(path, isFile, overwrite, permission, umask, isAppendBlob, eTag, tracingContext);
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:750]: any reason for the method. lets remove it?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 21]
```diff
@@ -605,6 +614,29 @@ public void deleteFilesystem(TracingContext tracingContext)
...
+  }
+
+  public AbfsRestOperation createPathBlob(final String path, final boolean isFile, final boolean overwrite,
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: I think its for test? lets have package-protected method.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java (hunk 14 above)
Comment (saxenapranav) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java:448]: In case some SAS rules changes for PutBlob and get differed from that of createPath. It would require, client to expose a new enum in future. So, suggesting maybe we have different enum from now, which will give leverage to dev to accomodate different rules for the PutBlob.

**Review by saxenapranav — COMMENTED** [review]
Looks good. Just nits. Probably would have to backmerge.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 20 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:750]: Sneha had a comment to add abstraction for these repetitive methods.

Comment (saxenapranav) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:750]: its one line command. How its helping?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 21 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: donr

**Review by saxenapranav — APPROVED** [review]
LGTM!
